### ISSUE TITLE
Questions in VATDecoder

### ISSUE THREAD

**Issue by jfzhuang** [yassouali/CCT#36]
Thanks for your excellent work and kindly code releasing. When delving deep into your codes, I found the implementation of the VATDecoder seems problematic. In the get_r_adv function, adv_distance is calculated by the kl_div between logp_hat and pred. However, logp_hat is the log_softmax result while pred is the softmax result, which are not compatible. In my understanding, the implementation should be like the following:
with torch.no_grad():
        logp = F.log_softmax(decoder(x_detached), dim=1)
for _ in range(it):
        logp_hat = F.log_softmax(decoder(x_detached + xi*d), dim=1)
        adv_distance = F.kl_div(logp_hat, logp, reduction='batchmean')
I wonder if my understanding correct. Waiting for your reply.

**Comment by yassouali** [comment]
Hi @jfzhuang 

Thank you for your intrest;

This is just with the kl_div implementation in pytorch, instead of passing the two probabilities, we need to pass log prob and prob.
Check out the docs: [here](https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss)

![image](https://user-images.githubusercontent.com/45082182/117236415-eba02d00-ae28-11eb-902c-3c2546175c0a.png)
